### DISCUSSION TITLE
Can the streaming be used in Android ?

### DISCUSSION THREAD

**Discussion by usfaa444** [ggml-org/whisper.cpp#560] [Bindings]
So I was going through this awesome library, and by the way thank you very much for such a great work, I am sure it will help thousands of developers, including me.

So I want to use the streaming to be able to transcribe any speech detected from the microphone, I saw that the stream demo you gave in the video, you typed a command in the CLI, is it possible to have this functionality implemented in Android please?

**Comment by rpikus27** [comment]
I'm currently implementing this idea on aosp. Where do you want the text to go after it is transcribed? What is the user experience you are trying to build?

**Reply by usfaa444** [reply]
So what I think of is, I have an android app where multiple people can join a call conference, and I want that whenever someone speaks, that I get the transcribed text from his speech and display it on a TextView.
I don't think recording the audio and storing it in a file and then transcrib it will be efficient here, so I saw the streaming example video you showed and I was thinking if we can have something like the streaming example video on Android, this will help me achieve my goal.

**Reply by usfaa444** [reply]
So I also decided to try to implement it.
So I copied the stream.cpp, CMakeLists.txt, common.h and common-sdl.h, and I link it to the Android gradle using cmake.

But now in the common-sdl.h file, it there is a problem of "SDL.h and SDL_audio.h files not found"
Please could you help me with that.

I can say that I barely started working with linking C++ code with Android so I don't really know what I should be doing to get that fixed. I know that you said that the stream is using SDL2, and we need to install it, I did.

Now in the CMakeLists.txt file, do we need to specify the path of the installed SDL2 library so that it can find the SDL.h and the SDL_audio.h files?

**Reply by rjpikus** [reply]
https://stackoverflow.com/questions/59750830/how-to-use-my-prebuilt-c-static-shared-library-in-aosp-project
https://android.googlesource.com/platform/build/soong/+/master/README.md
These might help. Google is migrating to Soong away from make files. Have you tried doing just that, specifying the path of the installed SDL2 Lib so CMakeLists.txt can find SDL.h and SDL_audio.h? I do not know off the top of my head.

**Reply by rjpikus** [reply]
Did you figure this out? What are you doing to make the calls? 
https://developer.android.com/reference/android/telecom/package-summary

**Reply by usfaa444** [reply]
Hello @rjpikus no I did not, and sorry for not getting back to you, I totally forgot. I am using Pexip for the calls

**Reply by rjpikus** [reply]
public static final int CAPABILITY_ADHOC_CONFERENCE_CALLING 
https://developer.android.com/reference/android/telecom/PhoneAccount
maybe use this inside android app if you want conferencing still

**Comment by usfaa444** [comment]
So I decided to not use the stream for my app, instead I am recording audio of 10seconds and putting them in a queue where I am transcribing them one at a time.
The problem I encountered is the following. 
When I test on a mibile phone, like my Samsung galaxy s22 ultra, everything is fine, and i have every audio transcribed with 3 to 4 seconds. 
But on some other devices, when transcribing the first audio, the transcsibData funtion in the LibWhisper get stuck it doesn't return anything until I close the app (the app goes to the background) and then at this momebt, then transcrib result is returned and the other audio in the queue get transcribed as well with 3 to 4 seconds. 

So any idea why when the app is running it is not transcribing the audio, and wait until the app goes to the background before?

Also to make sure of the behavior, I installed tour demo app on the same device and it is thz same behavior when I click on the TranscribSample button, it reads the audio, and then it doesn't show anything, when I click the back button and the app goes to the background and I open the app again, then I see the transcribed text on the screen.

**Reply by rjpikus** [reply]
What is the benefit of parsable/parsed segments? No more streaming because you worked around the issue? How is this better
?
To answer your question: One idea is that the interface between the app and os is not right or the data object isn't perfect. You can set timing logs to see where it's lagging. A solution: may be to look into the app tapping into the os. Your app is put into a package (APK). Android Application Framework is the interface between your APK and os, providing APIs and methods that enable the APK and os to interact. It's all code
.

**Comment by rpikus27** [comment]
I don't know why, but you can put logs in between your method calls. I am
using android.telecom to make calls

**Reply by usfaa444** [reply]
Yeah I did put sole logs but some of them are not getting called. They are called only after the app goes to the background

**Comment by rpdrewes** [comment]
I use whisper.cpp from my de-googled Android phone daily for months now for dictation to SMS and email and web search. My preferred method is to use the Konele app as the voice input frontend which sends the audio to my own private server which is running a very simple python wrapper around whisper.cpp. Here is my project: https://github.com/rpdrewes/whisper-websocket-server

Another alternative is this port of whisper.cpp to run natively on Android. I have tested it and it also works well but it is not as fast as my server-based approach in my environment (but if you have a slower server and/or a faster phone it might be better for you). It uses Konele as a frontend for the input as well, I believe. See: https://github.com/alex-vt/WhisperInput
